src/compiler/checker.ts

```diff
@@ -3328,25 +3329,62 @@ namespace ts {
+                if (!baseSymbol || baseSymbol === unknownSymbol) {
+                    // pessimistically asssume that base type is not independent
+                    return false;
+                }
+                const baseType = <InterfaceType>getDeclaredTypeOfSymbol(baseSymbol);
+                if (baseType.thisType) {
+                    return false;
+                }
+            }
+            else if (classSymbol) {
+                Debug.assert((classSymbol.flags & SymbolFlags.Class) !== 0);
+                const baseTypes = getBaseTypes(<InterfaceType>getDeclaredTypeOfClassOrInterface(classSymbol));
+                if (baseTypes.length) {
+                    return (<InterfaceType>baseTypes[0]).thisType === undefined;
+                }
+            }
+            return true;
+        }
+
+        // Returns true if the class\interface given by the symbol is free of "this" references. Specifically, the result is
+        // true if the class\interface itself contains no references to "this" in its body, if all base types are classes\interfaces,
+        // and if none of the base classes\interfaces have a "this" type.
+        function isIndependentClassOrInterface(symbol: Symbol): boolean {
             for (const declaration of symbol.declarations) {
+                if (declaration.flags & NodeFlags.UsesThisTypeOrReference) {
+                    return false;
+                }
                 if (declaration.kind === SyntaxKind.InterfaceDeclaration) {
-                    if (declaration.flags & NodeFlags.ContainsThis) {
+                    if (!isIndependentBaseInterfaceList(getInterfaceBaseTypeNodes(<InterfaceDeclaration>declaration))) {
                         return false;
                     }
-                    const baseTypeNodes = getInterfaceBaseTypeNodes(<InterfaceDeclaration>declaration);
-                    if (baseTypeNodes) {
-                        for (const node of baseTypeNodes) {
-                            if (isSupportedExpressionWithTypeArguments(node)) {
-                                const baseSymbol = resolveEntityName(node.expression, SymbolFlags.Type, /*ignoreErrors*/ true);
-                                if (!baseSymbol || !(baseSymbol.flags & SymbolFlags.Interface) || getDeclaredTypeOfClassOrInterface(baseSymbol).thisType) {
-                                    return false;
-                                }
-                            }
-                        }
+                }
+                else if (isClassLike(declaration)) {
+                    const isIndependent = isIndependentHeritageClauseElement(getClassExtendsHeritageClauseElement(declaration), symbol) &&
+                        isIndependentBaseInterfaceList(getClassImplementsHeritageClauseElements(declaration));
+
+                    if (!isIndependent) {
+                        return false;

```

Comment:
I think it would be easier to read as 

```ts
const isIndependentExtends = ...
const isIndependentImplements = ...
return isIndependentExtends && isIndependentImplements;
```

(The passthrough `if (!isIndependent) return false;` is a little suspicious *because* it doesn't do anything special.)